### PR TITLE
Testing infra - stablize data folder usage and clean up

### PR DESCRIPTION
**Pull request by bleskes**
The plan for persistent node ids ( #17811 ) is to tie the node identity to a file stored in it's data folders. As such it becomes important that nodes in our testing infra have better affinity with their data folders and that their data folders are not cleaned underneath them. The first is important because we fix the random seed used for node id generation (for reproducibility) and allowing the same node to use two different data folders causes two separate nodes to have the same id, which prevents the cluster from forming. The second is important, for example, where a full cluster restart / single node restart need to maintain node identity and wiping the data folders at the wrong moment prevents this.

Concretely this PR does the following:
1) Remove previous attempts to have data folder per role using a prefix. This wasn't effective as it was using the data paths settings which are only used for part of the runs. An attempt to completely separate the paths via the home dir failed due to assumptions made by index custom path about node data folder ordinal uniqueness  (see #19076)
2) Change full cluster restarts to start up nodes in the same order their were first created in, only randomly swapping nodes with the same roles.
3) Change test cluster reset methods to first shutdown the unneeded nodes and then re-start the shared nodes that were shut down, so they'll reclaim their data folders.
4) Improve data folder wiping logic and make sure it wipes only folders of "offline" nodes.
5) Add some very basic tests

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/gateway/Gateway.java [hunk 1]
```diff
@@ -159,15 +157,6 @@ public void performStateRecovery(final GatewayStateRecoveredListener listener) t
         builder.metaData(metaDataBuilder);
         listener.onSuccess(builder.build());
     }
-    public void reset() throws Exception {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/gateway/Gateway.java]: YES!

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 2]
```diff
@@ -1115,6 +1128,22 @@ private void wipeDataDirectories() {
         }
     }
 
+    private void markNodeDataDirsAsPendingForWipe(Node node) {
+        assert Thread.holdsLock(this);
+        NodeEnvironment nodeEnv = getInstanceFromNode(NodeEnvironment.class, node);
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: while we are at it maybe add a Node.getNodeEnvironment() method that doesn't go through guice?

**Comment by s1monw** [comment]
nice, left a small comment. I think eventually we want data dirs that should be reused to be explicit. but that can be another change

**Comment by bleskes** [comment]
thanks @s1monw I pushed a change with your request. tass pass and I merged it in.

**Pull request merged**